enterprise/server/remote_execution/nbd/nbdclient/nbdclient.go

```diff
@@ -0,0 +1,306 @@
+// according to the kernel config. "Creating" the device here just means that we
+// are activating one of these existing devices, connecting it to this
+// ClientDevice instance.
+func (d *ClientDevice) createDevice() error {
+	start := time.Now()
+
+	// Using background context here since cancelling the context will
+	// disconnect the device, and it needs to stay connected for the lifetime of
+	// the VM or until we explicitly disconnect it (for hot-swapping).
+	ctx := context.Background()
+
+	if err := d.updateMetadata(ctx); err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	idx, wait, err := nbd.Loopback(ctx, d, uint64(d.metadata.GetSizeBytes()))
+	if err != nil {
+		cancel()
+		return status.WrapErrorf(err, "failed to create device %q", d.metadata.GetName())
+	}
+	d.device = &device{
+		Index:  idx,
+		cancel: cancel,
+		wait:   wait,
+	}
+	if err := d.waitForReady(ctx); err != nil {
+		return status.WrapError(err, "ready check poll failed")
+	}
+
+	log.Infof("Created network block device %q at %s in %s", d.metadata.GetName(), d.DevicePath(), time.Since(start))
+	return nil
+}
+
+// waitForReady waits for the NBD device to become ready.
+func (d *ClientDevice) waitForReady(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, readyCheckTimeout)
+	defer cancel()
+
+	f, err := os.Open(fmt.Sprintf("/sys/block/%s/size", filepath.Base(d.DevicePath())))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	for {
+		if _, err := f.Seek(0, io.SeekStart); err != nil {
+			return err
+		}
+		rsize, err := io.ReadAll(f)
+		if err != nil {
+			return err
+		}
+		size, err := strconv.ParseInt(strings.TrimSpace(string(rsize)), 10, 64)
+		if err != nil {
+			return err
+		}
+		if size > 0 {
+			return nil
+		}
+		select {
+		case <-time.After(readyCheckPollInterval):
+			continue
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// Mount connects the NBD device if not already connected, updates the device
+// size parameter to match its current backing file size (according to the host
+// server), and mounts the client device to the given directory path. The
+// directory must already exist.
+func (d *ClientDevice) Mount(path string) error {
+	if d.mountPath != "" {
+		return status.FailedPreconditionError("already mounted")
+	}
+	d.mountPath = path
+	if d.device == nil {
+		if err := d.createDevice(); err != nil {
+			return status.WrapError(err, "failed to create device")
+		}
+	}
+	fstype := filesystemTypeString(d.metadata.GetFilesystemType())
+	if err := syscall.Mount(d.DevicePath(), path, fstype, syscall.MS_RELATIME, "" /*=data*/); err != nil {
+		return status.InternalErrorf("failed to mount %s (type %s) to %s: %s", d.DevicePath(), fstype, path, err)
+	}
+	log.Infof("Mounted %s (type %s) to %s", d.DevicePath(), d.metadata.GetName(), path)
+	return nil
+}
+
+// Unmount unmounts the device from its current mount path. It does not
+// disconnect the NBD device.
```

Comment:
It does disconnect the device right?

Reply:
haha woops, yeah this is an outdated comment